crates/solver/src/settlement_ranker.rs

```diff
@@ -164,11 +164,33 @@ impl SettlementRanker {
             );
         }
 
-        let (mut rated_settlements, errors) = self
+        let (rated_settlements, errors) = self
             .settlement_rater
             .rate_settlements(solver_settlements, external_prices, gas_price)
             .await?;
 
+        // Discard solutions with negative obj value
+        let mut rated_settlements = rated_settlements
+            .into_iter()
+            .filter(|(solver, rated_settlement, _)| {
+                // TODO for now, apply only for 0x solver
+                if solver.name() != "0x" {
+                    return true;
+                }
```

Comment:
This feels like a code smell - where we add solver-specific logic to the driver run loop.

Can we instead implement something on the solver level for this? A "DiscardNegativeObjectiveValueSolver" that only wraps the 0x solver?

Reply:
I wouldn't even create a wrapper. This should be done directly in the 0x code.

Reply:
This is indeed a red flag, however acceptable (IMO) since we will either:
1. remove this line and therefore apply the code change to all solvers
2. revert this PR completely.

That said, would you be ok with merging this as is?

Reply:
I would like to have it directly in the 0x solver code and only there. Unless it turns out that doing that has other problems.

Reply:
I don't think we have a settlement simulation / gas price and everything else needed to compute the objective value in the solver code itself.

Reply:
Our Solver trait contains all that is needed to calculate objective value afaik.

Reply:
> This should be done directly in the 0x code.

I agree.

Originally, I had not suggested this as 0x is implemented as a single order solver though, so we would need wrap it I think:
```
struct ZeroExSolver(SingleOrderSolver<Inner>);
```

The reason being that merging solutions may be a net benefit (i.e. push a part of a solution from having a negative objective value to a positive one).

Reply:
It can still be done directly in the zeroex solver implementation at `crates/solver/src/solver/zeroex_solver.rs`. Not super clean but you can convert a single settlement into a full settlement and calculate the objective value. I would then not emit any single settlement that has negative object value. It shouldn't be part of the merging.

Reply:
The issue is gas metering on Ethereum means that the `... - gas_estimate` part can change depending on whether or not the partial settlement is alone or part of a larger merged settlement (and potentially make the objective become positive).

I.e. given:
- Settlement A: -X objective
- Settlement B: +Y objective

It is possible that:
- Settlement A+B: +Z objective
- where Z > Y

Reply:
And, to me at least, this isn't super contrived, and related to our value proposition (batching trades is more efficient then settling them on their own, and can push us into economical viability).

Reply:
> I.e. given:
> 
>     * Settlement A: -X objective
> 
>     * Settlement B: +Y objective
> 
> 
> It is possible that:
> 
>     * Settlement A+B: +Z objective
> 
>     * where Z > Y

Interesting. Wasn't aware that this could happen! Could you briefly explain this or give a pointer so that I get an understanding of it?

Reply:
Specifically, on Ethereum, reusing state (accessing contract code and storage for example) comes at a reduced cost the second time around. Additionally, we have a fixed overhead per settlement for executing it on-chain (i.e. calling the settlement contract).

With this in mind, If we consider a recent solution that was executed at 15 Gwei, and that the fixed overhead of a settlement is ~90K gas, that is about -0.00135Ξ to the objective value that will not be there when the settlements are combined (about $1.60). This is even before we start considering the significant discounts for accessing the same contract twice within a transaction (which is 2500 gas per state access - be it a storage slot or contract code).